control-flow/index.md

```diff
@@ -1,6 +1,368 @@
 ---
-layout: default.njk
+layout: default
 title: Control Flow
 ---
 
-I've seen things you people wouldn't believe. Attack ships on fire off the shoulder of Orion. I watched C-beams glitter in the dark near the Tannhäuser Gate. All those moments will be lost in time, like control flow.
+As we create programs we write sequences of events. Usually, we expect them execute in order but this is not always the way it occurs. Even though JavaScript ios [single-threaded](https://medium.com/better-programming/is-node-js-really-single-threaded-7ea59bcc8d64 "Is Node.js Really Single-Threaded?"), functions that call APIs, setTimeouts and other structures can result in our code running in an unexpected order.
+
+Consider the following code.
+
+```
+console.log('Clean teeth');
+setTimeout(() => {
+  console.log('Use lavatory');
+});
+console.log('Wash face and hands');
+```
+
+In the above code our compiler goes through each line executing each instruction. First it logs `Clean teeth`. Then it doesn't hang around for the `setTimeout`, it simply carries on and logs `Wash face and hands`. Then (a small moment later), it will log `Use lavatory`.
+
+This is not what we want. We'll have to go wash our hands again. In production code, even worse things can happen.
+
+Thankfully Node offers 3 approaches we can use to control the order in which our code executes. _Callbacks_, _promises_ and _async/await_. Let's take a look at each.
+
+## Callbacks
+
+Earlier in Node's development, it was common to use the _callback pattern_ to control the order that code is executed. The pattern involves passing a function as a parameter (_callback_) into your functions, and then calling that _callback_ when you're ready to continue.
+
+Let's say we have to get some data from an external service. We don't know how long this service might take to respond. We would need to wait until the service responds, we can't simply continue running code as we might need the response.
+
+To simulate this I've created a `randomDelayedResponse` function that will return a response after an unspecified amount of time.
+
+```
+function randomDelayedResponse(text) {
+  // Using a timeout here for the sake of simulating an external request
+  const timeOut = Math.floor(Math.random() * 10) + 1;
+  const output = text;
+  setTimeout(() => {
+    return output;
+  }, timeOut * 10);
+}
+
+const output = randomDelayedResponse('Hello');
+console.log(output); // empty
+```
+
+Notice that the final line above returns `undefined`.
```

Comment:
Worth adding `This is because nothing is returned from the `randomDelayedResponse` function.`?